### PR TITLE
Completions are working again - ready for review

### PR DESCRIPTION
**Pull request by steveyegge**
This branch gets completions working again, by following the new document synchronization closely, and fixing a handful of bugs that crept in as the protocol changed.

The branch includes an integration test suite for document synchronization testing that mirrors the new IntelliJ test suite for the same.

### CODE REVIEW

File: CONTRIBUTING.org
Comment (dominiccooney) [inline CONTRIBUTING.org:37]: should consider ... at least ... to be ... for

You have blown my stack for dependent phrases. Be direct:

```suggestion
its responses in various ways. To get the best results, the WASM files for the
languages you use regularly are essential.
```

File: README.org
Comment (dominiccooney) [inline README.org:66]: ???

File: cody-repo-util.el
Comment (dominiccooney) [inline cody-repo-util.el:44]: Windows uses UNC paths too, like `\\fileserver\c$\foo\bar.txt`

So those should be `file://fileserver/c$/foo/bar.txt` or something like that, not `file:////...` although it might work.

Why are you messing with the case of the drive letter? Does it matter?

Your `"\\([a-zA-Z]\\):"` regexp is going to match URLs with ports, but I guess that doesn't happen in practice?

File: cody-repo-util.el
Comment (dominiccooney) [inline cody-repo-util.el:62]: When is `file-relative-name` ever 0 length? Outside the repo it's going to be some number of `../` and `file`; inside the directory it is going to be `file`, except if `file` is `""` and then we have `.`; etc.

File: cody-repo-util.el
Comment (dominiccooney) [inline cody-repo-util.el:65]: Would love a unit test or some comments indicating what these paths are like, because I'm not following.

File: cody-repo-util.el
Comment (dominiccooney) [inline cody-repo-util.el:78]: Is this used?

File: cody-repo-util.el
Comment (dominiccooney) [inline cody-repo-util.el:95]: In the cody.el `defcustom` this is `cody-remote-url-replacements` note the  `cody-` versus `cody--`

File: cody-repo-util.el
Comment (dominiccooney) [inline cody-repo-util.el:146]: For the TypeScript extension I just ported something from Go which did this and it seemed to be long-lived. Though NOW I'm told that there's a GraphQL for converting clone URLs to codebase names and admins can create a mapping, so the truth might depend whether you want to pass one of these names to a remote (and most of the time what you have here will be right anyway...)

File: cody-repo-util.el
Comment (dominiccooney) [inline cody-repo-util.el:128]: Is this really a remote branch name? Doesn't this just do `git rev-parse HEAD`, which is a commit hash? And it seems like `vc-git-working-revision` doesn't care about its filename argument? (...at least with the one I have with emacs-28.2.)

File: cody.el
Comment (dominiccooney) [inline cody.el:106]:
```suggestion
  "Default branch name for the current project."
```

File: cody.el
Comment (dominiccooney) [inline cody.el:139]:
```suggestion
The remote Agent is typically started by an IDE such as VS Code,
```

File: cody.el
Comment (dominiccooney) [inline cody.el:160]:
```suggestion
the actual document in Emacs. Setting this custom variable to non-nil,
```

**Review by dominiccooney — APPROVED** [review]
Approving to unblock, some comments and questions inline. Looking forward to using this.

File: CONTRIBUTING.org
Comment (keegancsmith) [reply CONTRIBUTING.org:35]: org mode link syntax is `[[link][description]]`
```suggestion
Cody uses [[https://tree-sitter.github.io/tree-sitter/][tree-sitter]] for improving
```

File: Eask
Comment (keegancsmith) [inline Eask:1]: TIL Eask. Looks neat.

File: Eask
Comment (keegancsmith) [inline Eask:37]: FYI when you next want to yakshave your init.el, latest versions of emacs have pretty decent builtin package management which also support installing from git repo like straight does. I recently switched from straight to just using that. https://github.com/keegancsmith/dotfiles/commit/09833c0e2ac306bce472dfead7c095c213ef476c

File: cody-repo-util.el
Comment (keegancsmith) [reply cody-repo-util.el:146]: yeah most of the time this will be right. On sourcegraph server side we do have a way which will convert things in the sourcegraph name based on remote. Indeed admins can set up templates for how to generate names. eg some instances will leave off the hostname from repo name to make nicer urls.

**Review by keegancsmith — COMMENTED** [review]
lots of nice looking code, to be honest I stopped reading. I think the best I can do here is to actually check out the branch and test rather than review code. I likely won't do that today, but I'll give it a shot once you have responded to @dominiccooney's review.

**Comment by keegancsmith** [comment]
Getting this inside of `cody--connection-create-process` when I run `cody-login`. I am using the jsonrpc library that is shipped with emacs 29.3
```
Initializing Cody connection...
slot-missing: Invalid slot name: "#<jsonrpc-process-connection jsonrpc-process-connection-2f033e2>", :events-buffer-config
```

FYI this is what I am using for install cody from this branch using emac's new vc-package stuff :)

```
(use-package cody
  :vc (cody :url "https://github.com/sourcegraph/emacs-cody.git"
            :branch "stevey/completions"
            :main-file "cody.el")
  :commands (cody-login cody-restart cody-chat cody-mode)
  :bind (("C-M-n" . cody-completion-cycle-next-key-dispatch)
         ("C-M-p" . cody-completion-cycle-prev-key-dispatch)
         ("M-TAB" . cody-completion-accept-key-dispatch)
         ("C-M-g" . cody-quit-key-dispatch))
  :init
  (setq cody--sourcegraph-host "sourcegraph.sourcegraph.com")
  (setq cody--access-token (auth-source-pick-first-password
                            :host cody--sourcegraph-host
                            :require '(:secret :host)))
  (setopt cody-workspace-root "/home/keegan/src/github.com/sourcegraph/cody")
  ;(setenv "MOCK_RG_PATH" "/opt/homebrew/bin/rg") ; ripgrep
  :config
  (defalias 'cody-start 'cody-login))
```

**Comment by keegancsmith** [comment]
I deleted the init slot for `:events-buffer-config`. From reading the jsonrpc code leaving it out will just change the format of the logs slightly.

I then ran into a hurdle that the keytar thing committed to the repo is darwin binary. Given we are passing in the access-token at init time, I would suspect we don't need integration into OS specific keychain?

```
/home/keegan/src/github.com/keegancsmith/dotfiles/.emacs.d/elpa/cody/dist/keytar-KOAAH267.node: Mach-O 64-bit arm64 bundle, flags:<NOUNDEFS|DYLDLINK|TWOLEVEL|WEAK_DEFINES|BINDS_TO_WEAK>

[stderr] Error: /home/keegan/src/github.com/keegancsmith/dotfiles/.emacs.d/elpa/cody/dist/keytar-KOAAH267.node: invalid ELF header
[stderr]     at Module._extensions..node (node:internal/modules/cjs/loader:1465:18)
[stderr]     at Module.load (node:internal/modules/cjs/loader:1206:32)
[stderr]     at Module._load (node:internal/modules/cjs/loader:1022:12)
[stderr]     at Module.require (node:internal/modules/cjs/loader:1231:19)
[stderr]     at require (node:internal/modules/helpers:179:18)
[stderr]     at ../node_modules/.pnpm/keytar@7.9.0/node_modules/keytar/lib/keytar.js (/home/keegan/src/github.com/keegancsmith/dotfiles/.emacs.d/elpa/cody/dist/index.js:93675:19)
[stderr]     at __require (/home/keegan/src/github.com/keegancsmith/dotfiles/.emacs.d/elpa/cody/dist/index.js:13:50)
[stderr]     at src/cli/auth/secrets.ts (/home/keegan/src/github.com/keegancsmith/dotfiles/.emacs.d/elpa/cody/dist/index.js:99637:29)
[stderr]     at __init (/home/keegan/src/github.com/keegancsmith/dotfiles/.emacs.d/elpa/cody/dist/index.js:10:56)
[stderr]     at src/cli/auth/AuthenticatedAccount.ts (/home/keegan/src/github.com/keegancsmith/dotfiles/.emacs.d/elpa/cody/dist/index.js:99755:5) {
[stderr]   code: 'ERR_DLOPEN_FAILED'
[stderr] }
[stderr] 
[stderr] Node.js v20.12.2
[internal] Tue Jul  2 21:36:25 2024:
(:message "Connection state changed" :change "exited abnormally with code 1\n")
```

File: CONTRIBUTING.org
Comment (steveyegge) [reply CONTRIBUTING.org:32]: Clarified, and also took a closer look at the whole file.

File: CONTRIBUTING.org
Comment (steveyegge) [reply CONTRIBUTING.org:37]: Fixed all the links.

File: CONTRIBUTING.org
Comment (steveyegge) [reply CONTRIBUTING.org:39]: fixed

File: README.org
Comment (steveyegge) [reply README.org]: fixed

File: cody-repo-util.el
Comment (steveyegge) [reply cody-repo-util.el:146]: I'd be happy to make a graphql call to get this correctly - will take a look.

But maybe after we merge!

File: cody-repo-util.el
Comment (steveyegge) [reply cody-repo-util.el:128]: fixed

File: cody-repo-util.el
Comment (steveyegge) [reply cody-repo-util.el:44]: This was an attempt at a faithful port of the JetBrains function [ProtocolTextDocument#uriFor](https://sourcegraph.com/github.com/sourcegraph/jetbrains@main/-/blob/src/main/kotlin/com/sourcegraph/cody/agent/protocol/ProtocolTextDocument.kt?L186:9).

File: cody-repo-util.el
Comment (steveyegge) [reply cody-repo-util.el:65]: I will add tests for all this. Much of it wound up not being used yet.

**Pull request merged**
